### ISSUE TITLE
Work with non-rooted CCTG

### ISSUE THREAD

**Issue by corneliusroemer** [mh-/corona-warn-companion-android#101]
In the last few days, a complete open source version of the CWA with MicroG as drop in replacement for Play Services ENF was launched.

It should be possible to use keys collected from MicroG directly, without need for a rooted phone. That'd make the app relevant for a whole lot more people.

Implementation shouldn't be too difficult, since MicroG is trying to be as close to Play services as possible. I could imagine you just have to switch out a path.

Would be amazing if you could do this 🙈

**Comment by mh-** [comment]
@corneliusroemer pls. see #99

**Comment by corneliusroemer** [comment]
Sweet! I realised that MicroG isn't the way to go for people who don't want to abandon Play services, afaik you can only use either Google or MicroG.

What's the status with #99, is this PR part of the current play store version, yet?

Will it be a separate mode (CCFG) or be worth of MicroG mode? Currently O don't have a CCFG mode, yet.

**Comment by mh-** [comment]
Can you please test https://github.com/mh-/corona-warn-companion-android/releases/tag/2.0.4 ? I don't think it will work without root, though, since it needs to pull the database from another app's file system area. I believe #97 would be required to be able to work without root, or CCTG storing the database on the SD card, like RaMBLE does.

**Comment by corneliusroemer** [comment]
@mh- Yes, you're right, it doesn't work. Didn't think it through, enough.

Basically there are two MicroG versions out there:
- a. the one bundled with CCTG if MicroG is not installed (this is what I do). MicroG uses the same namespace as play services to work as drop-in replacement. Thus its MicroG XOR Play Services and I'm not willing to go all in for MicroG - so I use the bundled ENF part of MicroG that comes with CCTG
- b. going all in with MicroG, then CCTG uses that version (apparently).

I think the problem with non-rooted phones is that anything in the /data folder can't be accessed by other apps.

Manual workaround may be to use adb and backup: https://stackoverflow.com/questions/13006315/how-to-access-data-data-folder-in-android-device

So you have to either broadcast the data (which is what @tlambertz seems to try to do) or you save it in a public folder, like /downloads (this is what Ramble seems todo). Either could work, though the export to downloads may be the easiest to implement.

If I'm not mistaken, it may be enough for CCTG to write the keys to download for your companion to read them - not sure whether they're interested in this though.

Am I making a mistake in my thinking anywhere? I'll try to get the keys through adb from my non-rooted phone, have you tried this?

**Comment by tlambertz** [comment]
You are correct. The `/data` directory contents are only readable for the owning app (at least on newer android versions).

It seems CCTG does have auto-backups enabled (which would usually lead to google backing up all app-data unencrypted, but as the target for this app is devices without google this is a sane choice). So I assume the adb-backup way would work. As this forces the use of an external computer it is somewhat annoying though.

As a matter of principle I feel like private keys should not be stored in a world-readable directory. Therefore I went for the export approach as outlined in #97 and https://github.com/microg/GmsCore/pull/1291. I should get around to fixing this so it can be merged.

CCTG seems to use upstream MicroG, and looking at [CCTG/#5](https://codeberg.org/corona-contact-tracing-germany/cwa-android/issues/5) it seems that you can simply open the MicroG Settings window, which includes the export button. After these two pull requests are merged and CCTG uses the new MicroG thinks might just work.

**Comment by corneliusroemer** [comment]
@tlambertz Thanks for your explanation! Just tried the adb-backup thing, nope, didn't work for me. Backup is empty.

I checked the Manifest and couldn't find that backup was disabled. Maybe backup is no longer supported in Android 11. Adb said backup was being deprecated. Oh well, we'll have to wait, I guess. Or I finally have to get round to rooting one of my old phones.

**Comment by djmaze** [comment]
@corneliusroemer I am using another very dirty workaround until export and import is working. (But you need to be able to use root via adb.)

In Android Studio, I adjusted the CWCA code and replaced the microG data directory in the `gmsPathStr` in *MicroGDbOnDisk.java* with the CWCA data directory. Also, I replaced the `sudo` command in *Sudo.java* with `bash`. So the app will in MicroG mode now take the db from its own data directory and also not try to use sudo. 

For getting the db into the right place, I am using `adb root` and `adb shell` to then `cp` the exposure.db from the the microG data directory to the CWCA data directory. Afterwards, I `chown` the db to the CWCA app user.

Really dirty, but works for me (on a non-rooted LineageOS with Android 10.).. Will now stay silent and make way for the clean solution :)

**Comment by corneliusroemer** [comment]
@djmaze Unfortunately `adb root` works only on debug builds, thus not for unrooted stock Android.
>By design adb root command works in development builds only (i.e. eng and userdebug which have ro.debuggable=1 by default).

**Comment by mh-** [comment]
merged #97, thanks to @tlambertz

**Comment by mh-** [comment]
Released as 2.0.5 https://github.com/mh-/corona-warn-companion-android/releases/tag/2.0.5 - please test it :)

**Comment by corneliusroemer** [comment]
@mh- 2.0.5 installs but no changes visible for me. What exactly should work now?
CCTG doesn't have an "export" function at the moment (is that coming with v1.8?).

**Comment by ale5000-git** [comment]
Export was just added to microG code today (not released yet) so until CCTG update to latest microG code there won't be nothing.

**Comment by tlambertz** [comment]
You will have to wait until microg publishes an update, then increase the version of microg in the `Corona-Warn-App/build.gradle` file in the CCTG repository accordingly

**Comment by ale5000-git** [comment]
New version of microG has been released.

**Comment by corneliusroemer** [comment]
> You will have to wait until microg publishes an update, then increase the version of microg in the `Corona-Warn-App/build.gradle` file in the CCTG repository accordingly

I updated the version in build.gradle, built it locally and installed on my phone.

@tlambertz  I now have access to the MicroG ENF screen, which shows me how many IDs I've collected. Where can I export the keys? I couldn't find a button  https://github.com/microg/GmsCore/pull/1291

ENF Version is 204215000, is that how it should be? Is it CCTG that's causing a problem here by not being totally up to date?
